Build a controller to upload container logs

stevekuznetsov opened the pull request:
Using the Kubernetes apimachinery clients and the client-go controllers
and work-queue, we can build a controller that acts on containers
terminating that will upload their logs to the appropriate place in GCS
so that we can expose the infrastructure-level logs that are necessary
to debug Prow pod decoration issues in failed jobs.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/area prow
/kind feature
/cc @BenTheElder @fejta 
/assign @cjwagner @kargakis

deads2k commented on prow/artifact-uploader/controller.go:
```diff
@@ -0,0 +1,198 @@
+		req, _ := labels.NewRequirement(kube.ProwJobIDLabel, selection.Exists, []string{})
+		options.LabelSelector = req.String()
+	}
+	podListWatcher := cache.NewFilteredListWatchFromClient(client.RESTClient(), "pods", api.NamespaceAll, optionsModifier)
```
`client.CoreV1().RESTClient()` perhaps?

stevekuznetsov commented:
Tested and working on our cluster. I'll break this PR up into smaller ones per-commit.

stevekuznetsov commented:
Rebased on top of #7920 #7919 #7917 #7914

stevekuznetsov commented:
/cc @BenTheElder @fejta
/assign @cjwagner @kargakis

kargakis commented on prow/artifact-uploader/controller.go:
```diff
@@ -0,0 +1,197 @@
+	core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
```
nit: wrong group

stevekuznetsov replied:
I wish GoLand would get this right -.-

kargakis commented on prow/artifact-uploader/controller.go:
```diff
@@ -0,0 +1,197 @@
+	indexer, informer := cache.NewIndexerInformer(podListWatcher, &api.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old interface{}, new interface{}) {
+			oldPod := old.(*api.Pod)
+			newPod := new.(*api.Pod)
```
Check that the resource version is not the same to avoid unnecessary resyncs.

kargakis commented on prow/artifact-uploader/controller.go:
```diff
@@ -0,0 +1,197 @@
+		UpdateFunc: func(old interface{}, new interface{}) {
+			oldPod := old.(*api.Pod)
+			newPod := new.(*api.Pod)
+
```
Also do you want to act on the pod once it's marked for deletion? Elsewhere we avoid resyncing when newPod.DeletionTimestamp != nil.

stevekuznetsov replied:
If it's marked for deletion and we gracefully terminate the containers we should still get logs for them, right? If we ignored the pod once it was marked for deletion would we lose those logs?

kargakis replied:
For the most common case, sinker initiates the deletion and by that time you already have pushed the logs but you are probably right for voluntary evictions (node drains).

kargakis commented on prow/artifact-uploader/controller.go:
```diff
@@ -0,0 +1,197 @@
+	informer cache.Controller
+
+	client        core.CoreV1Interface
+	prowJobClient *kube.Client
```
Any need for a client vs a lister that uses cached prowjobs?

stevekuznetsov replied:
Are you suggesting I build a lister and a cache, or do we already have one?

stevekuznetsov replied:
(We can get a lister and cache nicely later when I refactor the ProwJobs and we generate clients for them using the CRD generators)

kargakis replied:
Agreed

kargakis commented on prow/cmd/artifact-uploader/main.go:
```diff
@@ -0,0 +1,166 @@
+// BindOptions binds flags to options
+func (o *Options) BindOptions(flags *flag.FlagSet) {
+	flags.IntVar(&o.NumWorkers, "num-workers", 25, "Number of threads to use for processing updates.")
+	flags.StringVar(&o.ProwJobNamespace, "prow-job-ns", "", "Namespace containing ProwJobs.")
```
We already store the prowjob namespace in the prow config, although it's still static.

stevekuznetsov replied:
Loading the entire config seemed like overkill for one field.

kargakis commented on prow/cmd/artifact-uploader/main.go:
```diff
@@ -0,0 +1,166 @@
+	)
+
+	if err := o.Run(); err != nil {
+		logrus.WithError(err).Fatal("Failed to push artifacts to GCS")
```
This error message needs to be something different. Feels more like a "Failed to run the artifacts uploader"

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -63,6 +63,12 @@ const (
+	// We also name pods after the ProwJob that spawned them but
+	// this allows for multiple resources to be linked to one
+	// ProwJob.
+	ProwJobIDLabel = "prow.k8s.io/id"
```
s/build-id/id/ for more clarity?

kargakis commented:
Any reason for a separate flag vs read the prow config for the prowjob namespace? Apart from that question, lgtm, rest of the comments are minor.

stevekuznetsov commented on Gopkg.lock:
```diff
@@ -811,7 +802,7 @@
 
 [[projects]]
-  branch = "master"
+  branch = "release-1.10"
```
I did? And you were the one to review it and merge it :)

stevekuznetsov commented:
Rebased

@kargakis addressed most comments that were clearly needed to change
 - did not load full Prow config for the namespace name only -- should we? One monolithic config is starting to feel weird and we need just one field.
 - did not change the label name since that went in a different PR and seems semantic only? Can do it if we feel strongly

@cjwagner @fejta @BenTheElder PTAL

kargakis commented:
Regarding the prow config, it's going to get much better once Sen starts
splitting out jobs into separate files.

On Fri, May 4, 2018, 19:11 Steve Kuznetsov <notifications@github.com> wrote:

> Rebased
>
> @kargakis <https://github.com/kargakis> addressed most comments that were
> clearly needed to change
>
>    - did not load full Prow config for the namespace name only -- should
>    we? One monolithic config is starting to feel weird and we need just one
>    field.
>    - did not change the label name since that went in a different PR and
>    seems semantic only? Can do it if we feel strongly
>
> @cjwagner <https://github.com/cjwagner> @fejta <https://github.com/fejta>
> @BenTheElder <https://github.com/BenTheElder> PTAL
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/7912#issuecomment-386666732>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFfzfvDxK-Cc-owF1U5Cg_vPgJ1djyks5tvIvHgaJpZM4Tte5m>
> .
>

fejta commented:
Does this obviate the need for hijacking the ENTRYPOINT of the test image?

stevekuznetsov commented:
@fejta not quite yet. However, it may in the future.

fejta commented:
APPROVE generally
Seems like kargakis has the details figure out (thanks!)

/uncc

Please add me back if you want more feedback

stevekuznetsov commented on prow/cmd/artifact-uploader/main.go:
```diff
@@ -0,0 +1,166 @@
+// for the cluster we're deploying to. We prefer to
+// use in-cluster configuration if possible, but will
+// fall back to using default rules otherwise.
+func loadClusterConfig() (*rest.Config, error) {
```
This tries in-cluster first, if that does not work it falls back to other methods (`$KUBECONFIG`, `~/.kube/config`, etc). We can make it explicit later, like you say, if we don't want the controller deployed on the cluster with the ProwJobs or test Pods.

cjwagner replied:
Ok, so this won't work with separate build clusters as is because the in-cluster config for the service cluster will be found?
When would we fail to find in-cluster config?

stevekuznetsov commented:
@kargakis is loading the entire config to get the namespace a blocker here?

kargakis commented:
> @kargakis is loading the entire config to get the namespace a blocker here?

no, that's fine. we can change later if we want

stevekuznetsov commented:
@cjwagner @fejta can we get a LGTM + approve here?

fejta commented:
@cjwagner any more thoughts here?

@stevekuznetsov assign me when/if you need an lgtm

stevekuznetsov commented:
@kargakis you can do a LGTM too :)

kargakis commented:
/lgtm

kargakis commented:
Please add a README in a follow-up
